### PR TITLE
[NGPDCP-2927] Slide action/Swiper component

### PR DESCRIPTION
**Pull request by kaiceresy**
## Description

**1) Changes**

Add Slide action/Swiper component for H5 project.

**2) Impact**

the `useDrag` hooks

**3) UI Preview**

image
image
image

## Reference (optional)

[NGPDCP-2927](https://comfortdelgrotaxi.atlassian.net/browse/NGPDCP-2927)

## Check list

- [x] 1. Did you start and verify your changes?
- [x] 2. Did you run build to make sure that your changes can be built successfully?
- [x] 3. No !important flag, inline style in css
- [x] 4. No boilerplate codes (1)
- [x] 5. No duplicate code that exist in common functions?
- [x] 6. All of defined variables should have default value, check null and undefined?
- [x] 7. Use meaningful name for variables, no suffix 1,2,... Describes reference information for easier to understand what's inside. (2)
- [x] 9. Unsubscribed all of subscribers and even listeners when you don't need them.

```
(1)
Boilerplate codes is duplicated multiple times a bunch of the same code. This should be a common function to reuse through your code or you can use generic class / methods or design pattern to avoid the Boilerplate codes.
```

```
(2)
Follow theo coding convention
NO acronym variable name:
WRONG: el, elm
RIGHT: element
Your code can be read as a sentence
```

[NGPDCP-2927]: https://comfortdelgrotaxi.atlassian.net/browse/NGPDCP-2927?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: packages/react-compass/src/slide-action/slide-action.styles.ts [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+export const StyledSlideBg = styled('div', {
+  position: 'absolute',
+  width: 'calc(100% - $4)',
+  height: 'calc(100% - $4)',
+  borderRadius: '$full',
+  opacity: 0.2,
+
+  variants: {
+    bgType: {
```
Comment (hoapn-fsoft) [inline packages/react-compass/src/slide-action/slide-action.styles.ts:90]:
```suggestion
    backgroundType: {
```

avoid abbreviations

File: packages/react-compass/src/slide-action/slide-action.styles.ts [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+      static: {
+        opacity: 0,
+      },
+      slide: {
+        // transform: 'translateX(-110%)',
+        width: '$12',
+      },
+    },
+    slideBg: {
```
Comment (hoapn-fsoft) [inline packages/react-compass/src/slide-action/slide-action.styles.ts:99]:
```suggestion
    slideBackground: {
```

avoid abbreviations

File: packages/react-compass/src/slide-action/slide-action.tsx [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+      bgType = 'slide',
+      slideBg = 'mono',
+      enableDragAfterSuccess = false,
+      children,
+
+      ...delegated
+    },
+    ref,
+  ) => {
+    const SlideRef = useDOMRef<HTMLDivElement>(ref)
```
Comment (hoapn-fsoft) [inline packages/react-compass/src/slide-action/slide-action.tsx:33]:
```suggestion
    const slideRef = useDOMRef<HTMLDivElement>(ref)
```

File: packages/react-compass/src/slide-action/slide-action.tsx [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+    const SlideBgRef = useRef<HTMLDivElement>(null)
```
Comment (hoapn-fsoft) [inline packages/react-compass/src/slide-action/slide-action.tsx:34]:
```suggestion
    const slideBgRef = useRef<HTMLDivElement>(null)
```

File: packages/react-compass/src/slide-action/slide-action.tsx [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+    const SlideLabelRef = useRef<HTMLDivElement>(null)
```
Comment (hoapn-fsoft) [inline packages/react-compass/src/slide-action/slide-action.tsx:35]:
```suggestion
    const slideLabelRef = useRef<HTMLDivElement>(null)
```

File: packages/react-compass/src/slide-action/slide-action.tsx [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+        if (!SlideBgRef.current) {
+          return
+        }
+
+        if (bgType === 'static') {
+          SlideBgRef.current.style.setProperty('opacity', opacity.toFixed(2))
+          return
+        }
+
+        // SlideBgRef.current.style.setProperty('transform', transform)
```
Comment (hoapn-fsoft) [inline packages/react-compass/src/slide-action/slide-action.tsx:51]: remove comment

File: packages/react-compass/src/slide-action/slide-action.tsx [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+        if (!SlideRef.current || !SlideLabelRef.current) {
+          return
+        }
+        const calcDistance = (maxSlideDistance - x) / maxSlideDistance
+        SlideLabelRef.current.style.setProperty(
+          'opacity',
+          (calcDistance - 0.1).toFixed(2),
+        )
+
+        const bgOpacity = x / maxSlideDistance - 0.4
```
Comment (hoapn-fsoft) [inline packages/react-compass/src/slide-action/slide-action.tsx:68]: create constant for 0.4 value

File: packages/react-compass/src/slide-action/slide-dragger.tsx [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+  css,
+  slideRef,
+  icon,
+  color,
+  onDrag,
+  onDragEnd,
+  disableDrag = false,
+}: SlideDraggerProps) {
+  const slideDragRef = useRef<HTMLButtonElement>(null)
+  const SlideDragElement = slideDragRef.current
```
Comment (hoapn-fsoft) [inline packages/react-compass/src/slide-action/slide-dragger.tsx:22]:
```suggestion
  const slideDragElement = slideDragRef.current
```

**Review by hoapn-fsoft — CHANGES_REQUESTED** [review]

**Comment by hoapn-fsoft** [comment]
Error type
image

**Review by hoapn-fsoft — APPROVED** [review]
LGTM

**Review by pieliedie — APPROVED** [review]

**Review by hoapn-fsoft — APPROVED** [review]

**Pull request merged**